### PR TITLE
Improve memory footprint

### PR DESCRIPTION
**Pull request by GuillaumeFavelier**
This PR decreases the memory footprint by cleaning the internal list of renderers. I would really appreciate your opinion on this @banesullivan , @akaszynski.

I benchmarked the changes of this branch by using the package `memory_profiler` with the following (specific) use-case:

```sh
$ mprof run ipython --gui=qt
In [1]: run script.py
In [2]: exit()
```

and the content of `script.py`:
```py
import pyvista as pv

for j in range(5):
    for i in range(4):
        p = pv.BackgroundPlotter()
        p.show()
        for k in range(5):
            p.add_mesh(pv.Sphere().subdivide(4))
    pv.close_all()
```

I got the following results with `mprof plot` (`master` in blue, `PR` in black):
image

Note: Some tests didn't work locally for me. I guess it's because they use the internal renderer after `close()` has been called. I think maybe even calling `clear()` after `close()` shouldn't be safe. Also the given script reproduces roughly an expected behaviour with `sphinx-gallery`.

### CODE REVIEW

**Comment by GuillaumeFavelier** [comment]
My bad, I realize I didn't comment the curves: with the blue curve, we notice a constant increase in memory consumption and with the black one we notice the 5 bumps corresponding to the sequence of 5 "plots" closed each time.

**Comment by banesullivan** [comment]
I'm curious why the memory still increases - don't we want it to increase until the first `j` iteration then be relatively constant for the remainder of the `j` loop (outer)?

**Comment by banesullivan** [comment]
We might want to use a tool that can breakdown the memory usage of every object

**Comment by GuillaumeFavelier** [comment]
Concerning your question about memory, I particularly like the answer of ekhumoro on [StackOverflow](
https://stackoverflow.com/questions/18199422/pyqt-memory-usage):

> To put it crudely:
The amount of memory available to a python application can grow during its lifetime, but it can never fully shrink back to where it started.
There is no way to force python to release unused memory back to the system.
So the best you can do is try to manage the amount by which the memory grows. One obvious way to do this is to delete unneeded objects, so that newly created objects can re-use the memory that has already been allocated.

+1 for better memory monitoring but I wouldn't know where to start

**Comment by banesullivan** [comment]
It still concerns me that the memory keeps growing and that the new objects being created in the inner loop don't take up the available memory from the last iteration's cleanup

**Comment by banesullivan** [comment]
Either way, what you've added is a huge help!!! And a step in the right direction

**Comment by banesullivan** [comment]
That plot just scared me a bit into thinking our memory usage problem is worse than I thought

**Comment by banesullivan** [comment]
@GuillaumeFavelier - could you pull the changes from the master branch next time you push?

File: pyvista/plotting/plotting.py [hunk 1]
```diff
@@ -2138,6 +2139,13 @@ def close(self):
             except BaseException:
                 pass
 
+        for renderer in self.renderers:
+            renderer.camera_position = None
```
Comment (banesullivan) [inline pyvista/plotting/plotting.py]: We do not want to clear this - quite often users want to access the last used camera position after the plotter was closed

**Comment by banesullivan** [comment]
It seems there are some lingering issues from clearing all the renderers from the plotter on close since its often assumed they are still present untile the object itself has be cleanup.

I wonder if it makes more sense to add the snippet below that you added in a `__del__` method so they are cleaned up when the `BasePlotter` is destoryed rather than on close:

```py
...
for renderer in self.renderers:
    renderer.camera_position = None
    renderer.RemoveAllViewProps()
    renderer._actors = {}
    del renderer
self.renderers = []
...
```

**Comment by GuillaumeFavelier** [comment]
I moved the snippet into `__del__()` but I didn't notice the memory improvement on my use-case (`master` in blue, this branch in black):
Figure_1

I pushed the modifications in case someone want to reproduce.

IIRC `__del__()` will be called when the object is garbage collected, we have no guarantee when that will happen and AFAIK we can't force it. I think it's a more pessimistic method to free the ressources compared to explicitly stating that we are done with an object.

May I suggest an API change? We could rename the current `clear()` function into `reset()` and ensure that the new `clear()` function actually close and clear the object.

Also, `pyvista.close_all()` already call `close()` and then `clear()` for each plotter. What is the intention behind this function? Is it more close + reset or close + clear?

**Comment by larsoner** [comment]
You can request garbage collection with import gc then gc.collect. we can add it in the reset_modules functionality in our use of Sphinx gallery in MNE and it will probably fix our problem

**Comment by GuillaumeFavelier** [comment]
By being even more aggresive in the cleaning, I obtain the following (`master` in blue, previous in red and local in black):
Figure_1

**Comment by banesullivan** [comment]
@GuillaumeFavelier, can you push the latest changes? I’d like to experiment with that last scanerio you shared.

The more aggressive cleanup looks most ideal/what we’d expect

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by GuillaumeFavelier** [comment]
I pushed my last results. I think there is definitely room for improvement but so far, I couldn't achieve less than ~750MB of memory consumption on my machine for this example. I expect that the tests will succeed since I didn't modify either `close()` or `clear()` but created `deep_clear()` instead.

**Comment by banesullivan** [comment]
Continued in #338

**Pull request closed without merging**